### ISSUE TITLE
elemental + bind(c) produces weird results.

### ISSUE THREAD

**Issue by Leporacanthicus** [flang-compiler/f18-llvm-project#1364]
I'm not entirely sure if this is an issue with the compiler "allowing this" or an issue with the compiler generating incorrect code.

Main fortran program called test_elemental.f90:
```
program p
  implicit none
  
  interface
     pure elemental function square(x) bind(c)
       import
       integer, value :: x
       integer :: square
     end function square
  end interface

  integer, dimension(5) :: arr = (/ 1, 2, 3, 4, 5 /)

  print *, square(arr)
end program p
```

with `square` defined in square.c like this:
```
int square(int x)
{
	return x * x;
}
```

compile with `clang -c square.c` and `flang-new test_elemental.f90 square.o -o test_elemental`

Running this will give some large, sometimes negative, numbers - the same number each time. Clearly it is calculating the square of an address.

Similar program written in all Fortran:
```
module elem_m
  implicit none
contains
    pure elemental function square(x)
    integer, value :: x
    integer :: square
    square = x * x
  end function square
end module elem_m

program p
  use elem_m
  implicit none

  integer, dimension(5) :: arr = (/ 1, 2, 3, 4, 5 /)

  print *, square(arr)
end program p
```
This produces the correct values, but it does appear to load a value from a `fir.ref<i32>`

Code generated by the second sample:

```
  func @_QMelem_mPsquare(%arg0: !fir.ref<i32>) -> i32 {
    %0 = fir.alloca i32 {bindc_name = "square", uniq_name = "_QMelem_mFsquareEsquare"}
    %1 = fir.load %arg0 : !fir.ref<i32>
    %2 = fir.load %arg0 : !fir.ref<i32>
    %3 = arith.muli %1, %2 : i32
    fir.store %3 to %0 : !fir.ref<i32>
    %4 = fir.load %0 : !fir.ref<i32>
    return %4 : i32
  }
```

This came up when I was trying to fix issue #918.

**Comment by Leporacanthicus** [comment]
If I remove the fix for #918, the following code fails to compile. With the fix for #918 that is using passing arguments by value, it works.

Here's an example to show that problem:
```
module elem_m
  implicit none
  interface
     pure function square(x) bind(c)
       import
       integer, value :: x
       integer :: square
     end function square
  end interface
     
contains
    pure elemental function square_elem(x)
    integer, value :: x
    integer :: square_elem
    square_elem = square(x)
  end function square_elem
end module elem_m

program p
  use elem_m
  implicit none

  integer, dimension(5) :: arr = (/ 1, 2, 3, 4, 5 /)

  print *, square_elem(arr)
end program p
```

**Comment by Leporacanthicus** [comment]
pmk confirmed on Slack that `VALUE` should be interpreted in the way described in this issue.

>  pmk  18 hours ago
> f18 appears to ignore the VALUE attribute when calling an interoperable procedure (and probably also when defining one).
> 
> Mats Petersson  17 hours ago
> Well, it's not completely ignoring it - there's code specifically hit when using the VALUE attribute, that isn't used when the VALUE attribute is absent. [In fact, it really helped me fix issue #918 to have the parser recognise the attribute, but I had to implement it differently than the elemental code, which is why I raised this question - I may not even have noticed that it wasn't working if it hadn't asserted due to missing implementation].
> My question, really, is if specifying VALUE on the function argument should have the same effect as the C++:
> int function_without_value(int &x) and int function_with_value(int x)?
> 
> pmk  17 hours ago
> Yes.

**Comment by jeanPerier** [comment]
Yes, 18.3.6 point 2, (4) about BIND(C) is not respected by lowering: "any scalar dummy argument with the VALUE attribute is interoperable with the corresponding formal parameter of the prototype".

We have to decide what to do out of the BIND(C) context (I am not sure this is decided in https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/docs/Calls.md). I would be in favor of doing the same as in the BIND(C) case if this is reasonable. One point where we may have to be careful here is with derived type scalars, passing them as SSA value may lead to issues with "big" structs in LLVM due to the related load/store, and I am not sure what are the C requirements for passing structs here.

@Leporacanthicus, are you working on a fix ?

**Comment by Leporacanthicus** [comment]
I am not working on a fix right now, but if nobody else has a quick fix, I'll look at it in a bit.

If someone else is feeling a need to fix, feel free... :)

Yeah, I'm aware of the big structs load/store issues. I believe the conventional solution is simply ot call memcpy when the struct size is bigger than some boundary.  This is the caller's responsibility, but yes, we may need to worry about the effects of that in some way. 

I can't see anything in the calling convention documentation that is contrary to what changing the elemental calls to take value arguments as values.

**Comment by Leporacanthicus** [comment]
I'm not quite sure what the right solution is here: I believe BIND(C) combined with ELEMENTAL is invalid. So there's no specific need to fix THAT.

However, it remains a problem that the fix that I provided a few days ago ONLY works for BIND(C), and `BaseAddressValueAttribute` is not supported for non-elemental subroutines and functions - this causes the compiler to crash.

**Comment by Leporacanthicus** [comment]
I'm working on a solution for the simple case that asserts in the compiler.
